addons/pkg-npm/jaxrs/src/main/java/org/commonjava/indy/pkg/npm/jaxrs/NPMContentAccessHandler.java

```diff
@@ -0,0 +1,424 @@
+                                    formatOkResponseWithEntity( content, acceptInfo.getRawAccept(), builderModifier );
+                        }
+                        catch ( final IndyWorkflowException e )
+                        {
+                            logger.error(
+                                    String.format( "Failed to render content listing: %s from: %s. Reason: %s", path,
+                                                   name, e.getMessage() ), e );
+                            response = formatResponse( e, builderModifier );
+                        }
+                    }
+                    else
+                    {
+                        logger.info( "RETURNING: retrieval of content: {}:{}", sk, path );
+                        // open the stream here to prevent deletion while waiting for the transfer back to the user to start...
+                        InputStream in = item.openInputStream( true, eventMetadata );
+                        final Response.ResponseBuilder builder = Response.ok( new TransferStreamingOutput( in ) );
+                        setInfoHeaders( builder, item, sk, path, true, contentController.getContentType( path ),
+                                        contentController.getHttpMetadata( item ) );
+                        if ( builderModifier != null )
+                        {
+                            builderModifier.accept( builder );
+                        }
+                        response = builder.build();
+                        // generating .http-metadata.json for npm group retrieve to resolve header requirements
+                        if ( eventMetadata.get( STORAGE_PATH ) != null && StoreType.group == st )
+                        {
+                            generateHttpMetadataHeaders( item, request, response );
+                        }
+                    }
+                }
+                finally
+                {
+                    if ( handleLocking )
+                    {
+                        item.unlock();
+                    }
+                }
+            }
+            catch ( final IOException | IndyWorkflowException e )
+            {
+                logger.error( String.format( "Failed to download artifact: %s from: %s. Reason: %s", path, name,
+                                             e.getMessage() ), e );
+                response = formatResponse( e, builderModifier );
+            }
+        }
+
+        logger.info( "RETURNING RESULT: {}:{}", sk, path );
+        return response;
+    }
+
+    private void generateNPMContentsFromTransfer( final Transfer transfer, final EventMetadata eventMetadata,
+                                                  Response response,
+                                                  final Consumer<Response.ResponseBuilder> builderModifier )
+    {
+        if ( transfer == null )
+        {
+            return;
+        }
+
+        OutputStream out1 = null;
```

Comment:
Can we make the variable names more descriptive in this method? It's hard to follow the logic here, especially when you don't know the package.json file format very well (like me).

Also, do we have a test case that handles this tarball content logic? I didn't see a package.json file in the ftests directory that contained anything in the attachments field.

Reply:
Seems this is not fixed in the new code.

Reply:
I’ve added a simple test to verify the tarball and version meta generation, I should say the ftests are not enough already although I’ve done the manual tests to ensure the correctness, we actually add many particulars than before, I should sort them out and add later on.